### PR TITLE
HBASE-26786 [hboss] Limit synchronization from hot path of HBoss APIs

### PR DESCRIPTION
**Pull request by ankitsinghal**
I was thinking about how to write a unit test for this change. but, I could not understand why these APIs were made synchronized in the first place. If someone can share the background or any idea, I'll probably cover that.

### CODE REVIEW

File: hbase-oss/src/main/java/org/apache/hadoop/hbase/oss/sync/ZKTreeLockManager.java [hunk 1]
```diff
@@ -196,7 +197,7 @@ protected void readUnlock(Path p) throws IOException {
   We need to protect this block against potential concurrent calls to close()
    */
   @Override
-  protected synchronized boolean writeLockAbove(Path p) throws IOException {
+  protected boolean writeLockAbove(Path p) throws IOException {
```
Comment (joshelser) [inline hbase-oss/src/main/java/org/apache/hadoop/hbase/oss/sync/ZKTreeLockManager.java:200]: HBASE-24961 #14 introduced the synchronized here. I think Wellington was fixing an issue where SecureBulkLoadManager was causing problems with the `FileSystem.CACHE`. That one was caused by HBASE-23679 which I had tried to fix. Need to make sure we have testing for bulk loading to make sure we don't regress.

File: hbase-oss/src/main/java/org/apache/hadoop/hbase/oss/sync/ZKTreeLockManager.java [hunk 2]
```diff
@@ -380,24 +382,23 @@ String describeLock(InterProcessReadWriteLock lock) {
...
 
-  public synchronized Map<Path,InterProcessReadWriteLock> getUnmodifiableCache() {
+  public Map<Path,InterProcessReadWriteLock> getUnmodifiableCache() {
     return Collections.unmodifiableMap(lockCache);
   }
 
-  private synchronized InterProcessReadWriteLock get(Path path) throws IOException {
-    if (!lockCache.containsKey(path)) {
+  private InterProcessReadWriteLock get(Path path) {
+    return lockCache.computeIfAbsent(path, s -> {
```
Comment (joshelser) [inline hbase-oss/src/main/java/org/apache/hadoop/hbase/oss/sync/ZKTreeLockManager.java:390]: This makes sense to me.

**Review by joshelser — COMMENTED** [review]
I'm a little worried about lessening the `writeLockAbove`. Sent Ankit some context to the original case that Wellington saw internally. I think the rest of this makes sense.

I would guess that we don't have coverage for bulk loading inside hbase-filesystem itself, so we'd have to do an end-to-end test by hand to verify we don't regress with this change.

**Review by joshelser — APPROVED** [review]
I'm trying to think of this in a wholistic way. This all stemmed from SecureBulkLoadManager having multiple threads, all referring to the same FileSystem instance. We let other threads pick up a FileSystem which had a closed Curator client.

We have to rely that SecureBulkLoadManager, in a multi-threaded environment, is closing a FileSystem object only when it's safe to do so and that no other clients can use it after that point.

If we assume the above to be true, then we do not need exclusion between writeLockAbove() and close() because we should know that, if the object is being close()'d, no one else is using it. The exclusion between close() and the rest of the calls shouldn't be required.

**Comment by ankitsinghal** [comment]
Thanks @joshelser  for taking a closer look.. I agree filesytem shouldn't need to worry about the synchronization between close and writes as it is what the caller should ensure.

**Pull request merged**